File: crates/valka-sdk/src/worker.rs
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

use futures::StreamExt;
use tokio::sync::mpsc;
use tokio_stream::wrappers::ReceiverStream;
use tonic::transport::Channel;
use tracing::{error, info, warn};
use uuid::Uuid;

use valka_proto::worker_service_client::WorkerServiceClient;
use valka_proto::*;
use valka_proto::{worker_request, worker_response};

use crate::context::TaskContext;
use crate::error::SdkError;
use crate::retry::RetryPolicy;

pub type TaskHandler = Arc<
    dyn Fn(TaskContext) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, String>> + Send>>
        + Send
        + Sync,
>;

/// Builder for creating a ValkaWorker.
pub struct ValkaWorkerBuilder {
    name: String,
    server_addr: String,
    queues: Vec<String>,
    concurrency: i32,
    handler: Option<TaskHandler>,
    metadata: String,
}

impl ValkaWorkerBuilder {
    pub fn new() -> Self {
        Self {
            name: format!("worker-{}", &Uuid::now_v7().to_string()[..8]),
            server_addr: "http://[::1]:50051".to_string(),
            queues: vec![],
            concurrency: 1,
            handler: None,
            metadata: String::new(),
        }
    }

    pub fn name(mut self, name: &str) -> Self {
        self.name = name.to_string();
        self
    }

    pub fn server_addr(mut self, addr: &str) -> Self {
        self.server_addr = addr.to_string();
        self
    }

    pub fn queues(mut self, queues: &[&str]) -> Self {
        self.queues = queues.iter().map(|s| s.to_string()).collect();
        self
    }

    pub fn concurrency(mut self, n: i32) -> Self {
        self.concurrency = n;
        self
    }

    pub fn handler<F, Fut>(mut self, f: F) -> Self
    where
        F: Fn(TaskContext) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = Result<serde_json::Value, String>> + Send + 'static,
    {
        self.handler = Some(Arc::new(move |ctx| Box::pin(f(ctx))));
        self
    }

    pub fn metadata(mut self, metadata: &str) -> Self {
        self.metadata = metadata.to_string();
        self
    }

    pub async fn build(self) -> Result<ValkaWorker, SdkError> {
        let handler = self
            .handler
            .ok_or_else(|| SdkError::Handler("No handler provided".to_string()))?;

        Ok(ValkaWorker {
            worker_id: Uuid::now_v7().to_string(),
            name: self.name,
            server_addr: self.server_addr,
            queues: self.queues,
            concurrency: self.concurrency,
            handler,
            metadata: self.metadata,
        })
    }
}

impl Default for ValkaWorkerBuilder {
    fn default() -> Self {
        Self::new()
    }
}

/// A Valka worker that connects to the control plane and processes tasks.
pub struct ValkaWorker {
    worker_id: String,
    name: String,
    server_addr: String,
    queues: Vec<String>,
    concurrency: i32,
    handler: TaskHandler,
    metadata: String,
}

impl ValkaWorker {
    pub fn builder() -> ValkaWorkerBuilder {
        ValkaWorkerBuilder::new()
    }

    /// Run the worker event loop. Blocks until shutdown.
    pub async fn run(self) -> Result<(), SdkError> {
        let mut retry_policy = RetryPolicy::new();

        loop {
            match self.connect_and_run(&mut retry_policy).await {
                Ok(()) => {
                    info!("Worker disconnected gracefully");
                    return Ok(());
                }
                Err(e) => {
                    let delay = retry_policy.next_delay();
                    warn!(
                        error = %e,
                        retry_in_ms = delay.as_millis(),
                        "Worker connection lost, reconnecting..."
                    );
                    tokio::time::sleep(delay).await;
                }
            }
        }
    }

    async fn connect_and_run(&self, retry_policy: &mut RetryPolicy) -> Result<(), SdkError> {
        let channel = Channel::from_shared(self.server_addr.clone())
            .map_err(|e| SdkError::Connection(e.to_string()))?
            .connect()
            .await?;

        let mut client = WorkerServiceClient::new(channel);

        // Set up bidirectional stream
        let (request_tx, request_rx) = mpsc::channel::<WorkerRequest>(256);
        let outbound = ReceiverStream::new(request_rx);

        let response = client.session(outbound).await?;
        let mut inbound = response.into_inner();

        retry_policy.reset();
        info!(worker_id = %self.worker_id, name = %self.name, "Connected to server");

        // Send hello
        let hello = WorkerRequest {
            request: Some(worker_request::Request::Hello(WorkerHello {
                worker_id: self.worker_id.clone(),
                worker_name: self.name.clone(),
                queues: self.queues.clone(),
                concurrency: self.concurrency,
                metadata: self.metadata.clone(),
            })),
        };
        request_tx
            .send(hello)
            .await
            .map_err(|_| SdkError::NotConnected)?;

        // Start heartbeat loop
        let hb_tx = request_tx.clone();
        let _worker_id = self.worker_id.clone();
        let hb_handle = tokio::spawn(async move {
            let mut interval = tokio::time::interval(tokio::time::Duration::from_secs(10));
            loop {
                interval.tick().await;
                let hb = WorkerRequest {
                    request: Some(worker_request::Request::Heartbeat(Heartbeat {
                        active_task_ids: vec![],
                        timestamp_ms: chrono::Utc::now().timestamp_millis(),
                    })),
                };
                if hb_tx.send(hb).await.is_err() {
                    break;
                }
            }
        });

        // Process incoming messages
        let semaphore = Arc::new(tokio::sync::Semaphore::new(self.concurrency as usize));

        loop {
            tokio::select! {
                msg = inbound.next() => {
                    match msg {
                        Some(Ok(response)) => {
                            match response.response {
                                Some(worker_response::Response::TaskAssignment(assignment)) => {
                                    let permit = semaphore.clone().acquire_owned().await
                                        .map_err(|_| SdkError::ShuttingDown)?;
                                    let handler = self.handler.clone();
                                    let tx = request_tx.clone();
                                    tokio::spawn(async move {
                                        let ctx = TaskContext::new(
                                            assignment.task_id.clone(),
                                            assignment.task_run_id.clone(),
                                            assignment.queue_name,
                                            assignment.task_name,
                                            assignment.attempt_number,
                                            assignment.input,
                                            assignment.metadata,
                                            tx.clone(),
                                        );

                                        let result = handler(ctx).await;

                                        let task_result = match result {
                                            Ok(output) => TaskResult {
                                                task_id: assignment.task_id,
                                                task_run_id: assignment.task_run_id,
                                                success: true,
                                                retryable: false,
                                                output: output.to_string(),
                                                error_message: String::new(),
                                            },
                                            Err(err) => TaskResult {
                                                task_id: assignment.task_id,
                                                task_run_id: assignment.task_run_id,
                                                success: false,
                                                retryable: true,
                                                output: String::new(),
                                                error_message: err,
                                            },
                                        };

                                        let request = WorkerRequest {
                                            request: Some(worker_request::Request::TaskResult(task_result)),
                                        };
                                        let _ = tx.send(request).await;
                                        drop(permit);
                                    });
                                }
                                Some(worker_response::Response::TaskCancellation(cancel)) => {
                                    info!(task_id = %cancel.task_id, "Task cancelled by server");
                                }
                                Some(worker_response::Response::HeartbeatAck(_)) => {}
                                Some(worker_response::Response::ServerShutdown(shutdown)) => {
                                    info!(reason = %shutdown.reason, "Server shutting down");
                                    break;
                                }
                                None => {}
                            }
                        }
                        Some(Err(e)) => {
                            error!(error = %e, "Stream error");
                            break;
                        }
                        None => {
                            info!("Server closed stream");
                            break;
                        }
                    }
                }
                _ = tokio::signal::ctrl_c() => {
                    info!("SIGINT received, shutting down gracefully");
                    let shutdown = WorkerRequest {
                        request: Some(worker_request::Request::Shutdown(GracefulShutdown {
                            reason: "SIGINT".to_string(),
                        })),
                    };
                    let _ = request_tx.send(shutdown).await;
                    // Wait for in-flight tasks
                    let _ = semaphore.acquire_many(self.concurrency as u32).await;
                    hb_handle.abort();
                    return Ok(());
                }
            }
        }

        hb_handle.abort();
        Err(SdkError::Connection("Stream closed".to_string()))
    }
}

File: sdks/python/src/valka/worker.py
"""Valka gRPC worker with bidirectional streaming."""

from __future__ import annotations

import asyncio
import json
import logging
import signal
import time
import uuid
from typing import Any, Awaitable, Callable

import grpc
import grpc.aio

from valka.context import TaskContext
from valka.errors import ConnectionError, HandlerError
from valka.retry import RetryPolicy

logger = logging.getLogger("valka.worker")

TaskHandler = Callable[[TaskContext], Awaitable[Any]]


class ValkaWorker:
    """gRPC bidirectional streaming worker for processing tasks.

    Use the builder pattern::

        worker = (
            ValkaWorker.builder()
            .name("my-worker")
            .server_addr("localhost:50051")
            .queues(["emails", "notifications"])
            .concurrency(4)
            .handler(my_handler)
            .build()
        )
        await worker.run()
    """

    def __init__(
        self,
        *,
        worker_id: str,
        name: str,
        server_addr: str,
        queues: list[str],
        concurrency: int,
        metadata: dict[str, Any] | None,
        handler: TaskHandler,
    ) -> None:
        self._worker_id = worker_id
        self._name = name
        self._server_addr = server_addr
        self._queues = queues
        self._concurrency = concurrency
        self._metadata = metadata
        self._handler = handler

        self._semaphore = asyncio.Semaphore(concurrency)
        self._active_tasks: dict[str, asyncio.Task[None]] = {}
        self._shutting_down = False
        self._shutdown_event = asyncio.Event()
        self._stream: grpc.aio.StreamStreamCall | None = None  # type: ignore[type-arg]

    @staticmethod
    def builder() -> ValkaWorkerBuilder:
        """Create a new worker builder."""
        return ValkaWorkerBuilder()

    @staticmethod
    def create(
        *,
        name: str = "",
        server_addr: str = "localhost:50051",
        queues: list[str],
        concurrency: int = 1,
        metadata: dict[str, Any] | None = None,
        handler: TaskHandler,
    ) -> ValkaWorker:
        """Create a worker directly without builder."""
        worker_id = str(uuid.uuid4())
        return ValkaWorker(
            worker_id=worker_id,
            name=name or f"python-worker-{worker_id[:8]}",
            server_addr=server_addr,
            queues=queues,
            concurrency=concurrency,
            metadata=metadata,
            handler=handler,
        )

    async def run(self) -> None:
        """Start the worker. Blocks until shutdown or unrecoverable error."""
        loop = asyncio.get_running_loop()
        for sig in (signal.SIGINT, signal.SIGTERM):
            loop.add_signal_handler(sig, lambda: asyncio.ensure_future(self.shutdown()))

        retry = RetryPolicy()
        while not self._shutting_down:
            try:
                await self._session(retry)
            except Exception as exc:
                if self._shutting_down:
                    break
                delay = retry.next_delay_seconds()
                logger.warning("Connection lost (%s), reconnecting in %.1fs", exc, delay)
                await asyncio.sleep(delay)

    async def shutdown(self) -> None:
        """Initiate graceful shutdown."""
        if self._shutting_down:
            return
        self._shutting_down = True
        logger.info("Shutting down, draining %d active tasks...", len(self._active_tasks))

        # Send graceful shutdown message
        if self._stream is not None:
            try:
                from valka._proto.valka.v1 import worker_pb2

                request = worker_pb2.WorkerRequest(
                    graceful_shutdown=worker_pb2.GracefulShutdown(reason="client shutdown")
                )
                await self._stream.write(request)
            except Exception:
                pass

        # Wait up to 30s for active tasks to drain
        if self._active_tasks:
            tasks = list(self._active_tasks.values())
            try:
                await asyncio.wait_for(asyncio.gather(*tasks, return_exceptions=True), timeout=30)
            except asyncio.TimeoutError:
                logger.warning("Drain timeout, cancelling %d tasks", len(self._active_tasks))
                for task in self._active_tasks.values():
                    task.cancel()

        self._shutdown_event.set()

    async def _session(self, retry: RetryPolicy) -> None:
        from valka._proto.valka.v1 import worker_pb2, worker_pb2_grpc

        channel = grpc.aio.insecure_channel(
            self._server_addr,
            options=[
                ("grpc.keepalive_time_ms", 10_000),
                ("grpc.keepalive_timeout_ms", 5_000),
            ],
        )

        try:
            stub = worker_pb2_grpc.WorkerServiceStub(channel)
            self._stream = stub.Session()

            # Send hello
            meta_str = json.dumps(self._metadata) if self._metadata else ""
            hello = worker_pb2.WorkerRequest(
                hello=worker_pb2.WorkerHello(
                    worker_id=self._worker_id,
                    worker_name=self._name,
                    queues=self._queues,
                    concurrency=self._concurrency,
                    metadata=meta_str,
                )
            )
            await self._stream.write(hello)

            retry.reset()
            logger.info(
                "Connected as %s (id=%s, queues=%s, concurrency=%d)",
                self._name,
                self._worker_id,
                self._queues,
                self._concurrency,
            )

            # Start heartbeat task
            heartbeat_task = asyncio.create_task(self._heartbeat_loop())

            try:
                async for response in self._stream:
                    if self._shutting_down:
                        break
                    await self._handle_response(response)
            finally:
                heartbeat_task.cancel()
                try:
                    await heartbeat_task
                except asyncio.CancelledError:
                    pass
        finally:
            await channel.close()
            self._stream = None

    async def _heartbeat_loop(self) -> None:
        from valka._proto.valka.v1 import worker_pb2

        while not self._shutting_down:
            await asyncio.sleep(10)
            if self._stream is None:
                break
            try:
                heartbeat = worker_pb2.WorkerRequest(
                    heartbeat=worker_pb2.Heartbeat(
                        active_task_ids=list(self._active_tasks.keys()),
                        timestamp_ms=int(time.time() * 1000),
                    )
                )
                await self._stream.write(heartbeat)
            except Exception:
                break

    async def _handle_response(self, response: Any) -> None:
        kind = response.WhichOneof("message")
        if kind == "task_assignment":
            await self._handle_task_assignment(response.task_assignment)
        elif kind == "task_cancellation":
            self._handle_task_cancellation(response.task_cancellation)
        elif kind == "server_shutdown":
            logger.info("Server shutdown: %s", response.server_shutdown.reason)
            await self.shutdown()
        elif kind == "heartbeat_ack":
            pass

    async def _handle_task_assignment(self, assignment: Any) -> None:
        await self._semaphore.acquire()
        task = asyncio.create_task(self._execute_task(assignment))
        self._active_tasks[assignment.task_id] = task
        task.add_done_callback(lambda _t: self._task_done(assignment.task_id))

    def _task_done(self, task_id: str) -> None:
        self._active_tasks.pop(task_id, None)
        self._semaphore.release()

    async def _execute_task(self, assignment: Any) -> None:
        from valka._proto.valka.v1 import worker_pb2

        ctx = TaskContext(
            task_id=assignment.task_id,
            task_run_id=assignment.task_run_id,
            queue_name=assignment.queue_name,
            task_name=assignment.task_name,
            attempt_number=assignment.attempt_number,
            raw_input=assignment.input,
            raw_metadata=assignment.metadata,
            send_fn=self._send,
        )

        success = False
        retryable = True
        output = ""
        error_message = ""

        try:
            result = await self._handler(ctx)
            success = True
            if result is not None:
                output = json.dumps(result) if not isinstance(result, str) else result
        except HandlerError as exc:
            retryable = exc.retryable
            error_message = str(exc)
            logger.warning("Task %s handler error: %s", assignment.task_id, exc)
        except Exception as exc:
            error_message = str(exc)
            logger.warning("Task %s failed: %s", assignment.task_id, exc)

        result_msg = worker_pb2.WorkerRequest(
            task_result=worker_pb2.TaskResult(
                task_id=assignment.task_id,
                task_run_id=assignment.task_run_id,
                success=success,
                retryable=retryable,
                output=output,
                error_message=error_message,
            )
        )
        await self._send(result_msg)

    async def _send(self, request: Any) -> None:
        if self._stream is not None:
            await self._stream.write(request)


class ValkaWorkerBuilder:
    """Fluent builder for ValkaWorker."""

    def __init__(self) -> None:
        self._name: str = ""
        self._server_addr: str = "localhost:50051"
        self._queues: list[str] = []
        self._concurrency: int = 1
        self._metadata: dict[str, Any] | None = None
        self._handler: TaskHandler | None = None

    def name(self, name: str) -> ValkaWorkerBuilder:
        """Set the worker display name."""
        self._name = name
        return self

    def server_addr(self, addr: str) -> ValkaWorkerBuilder:
        """Set the gRPC server address (host:port)."""
        self._server_addr = addr
        return self

    def queues(self, queues: list[str]) -> ValkaWorkerBuilder:
        """Set the queues this worker consumes from."""
        self._queues = queues
        return self

    def concurrency(self, n: int) -> ValkaWorkerBuilder:
        """Set max concurrent task handlers."""
        self._concurrency = n
        return self

    def metadata(self, meta: dict[str, Any]) -> ValkaWorkerBuilder:
        """Set optional worker metadata."""
        self._metadata = meta
        return self

    def handler(self, fn: TaskHandler) -> ValkaWorkerBuilder:
        """Set the task handler function."""
        self._handler = fn
        return self

    def build(self) -> ValkaWorker:
        """Build the worker. Raises ValueError if queues or handler missing."""
        if not self._queues:
            raise ValueError("At least one queue is required")
        if self._handler is None:
            raise ValueError("A handler function is required")

        worker_id = str(uuid.uuid4())
        return ValkaWorker(
            worker_id=worker_id,
            name=self._name or f"python-worker-{worker_id[:8]}",
            server_addr=self._server_addr,
            queues=self._queues,
            concurrency=self._concurrency,
            metadata=self._metadata,
            handler=self._handler,
        )
